src/EditorFeatures/Core/Diagnostics/AbstractPushOrPullDiagnosticsTaggerProvider.PullDiagnosticsTaggerProvider.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Workspaces;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Tagging;
+
+namespace Microsoft.CodeAnalysis.Diagnostics;
+
+internal abstract partial class AbstractPushOrPullDiagnosticsTaggerProvider<TTag> where TTag : ITag
+{
+    /// <summary>
+    /// Base type for all taggers that interact with the <see cref="IDiagnosticAnalyzerService"/> and produce tags for the
+    /// diagnostics with different UI presentations.  It does no computation work itself, but instead defers that to it's
+    /// underlying <see cref="SingleDiagnosticKindPullTaggerProvider"/>s.
+    /// </summary>
+    private sealed partial class PullDiagnosticsTaggerProvider : ITaggerProvider
```

Comment:
thisi s the same modern '4 way pull tagger' added earlier this week. The only difference is that it is now no longer abstract. It is no longer subclassed, but instead is passed the subclassable entity as it's first parameter to get the specific logic for each type of diagnostic tagging.

It was previously called AbstractAggregateDiagnosticTaggerProvider.